shell/server/core.js

```diff
@@ -332,13 +393,34 @@ restoreInternal = (tokenId, ownerPattern, requirements, parentToken) => {
       token.objectId.appRef = new Buffer(token.objectId.appRef);
     }
 
+    // Ensure the grain is running, then restore the capability.
     return waitPromise(globalBackend.useGrain(token.grainId, (supervisor) => {
       return supervisor.restore(token.objectId, requirements, parentToken);
     }));
   } else if (token.parentToken) {
+    // A token which chains to some parent token.  Restore the parent token (possibly recursively),
+    // checking requirements on the way up.
     return restoreInternal(token.parentToken, Match.Any, requirements, parentToken);
+  }
+
+  // Note that hereafter, `token` should be, by process of elimination, a UiView token.
+  if (!token.grainId) {
+    throw new Meteor.Error(500, "Expected token " + token._id + " to have a grainId");
+  }
+
+  if (token.owner && token.owner.grain) {
```

Comment:
I don't think this logic should be based on owner type. restore() should always return a "sealed" UiView, unless the token is explicitly marked as allowing direct access (a marking we have not yet defined). Human users opening the view in the sandstorm UI go through a completely different path that doesn't involve restore(), so we don't need to account for them here. So I would remove this if() and its whole else branch.